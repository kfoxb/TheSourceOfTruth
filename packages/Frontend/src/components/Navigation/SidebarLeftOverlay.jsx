import React, { Component } from 'react';
import { Sidebar, Menu } from 'semantic-ui-react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

export default class SidebarLeftOverlay extends Component {
  static propTypes = {
    isAuthenticated: PropTypes.bool.isRequired,
    logout: PropTypes.func.isRequired,
    sideBarVisibility: PropTypes.bool.isRequired,
    toggleMenu: PropTypes.func.isRequired,
  }

  componentDidMount() {
    document.addEventListener('mousedown', this.handleOffClick);
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleOffClick);
  }

  handleLogout = () => {
    this.props.logout();
    this.props.toggleMenu();
  };

  handleOffClick = (e) => {
    if (this.ref && !this.ref.contains(e.target) && this.props.sideBarVisibility) {
      this.props.toggleMenu();
    }
  };

  handleRef = (el) => { this.ref = el; }

  renderLogout = () => (
    <Menu.Item
      as="div"
      name="logout"
      onClick={this.handleLogout}
    >
      Sign Out
    </Menu.Item>
  )


  renderMenuItem = (name, label, path) => (
    <Link href={path} to={path}>
      <Menu.Item as="div" name={name} onClick={this.props.toggleMenu}>
        {label}
      </Menu.Item>
    </Link>
  );

  render() {
    const { renderMenuItem } = this;
    return (
      <div ref={this.handleRef}>
        <Sidebar as={Menu} animation="overlay" width="thin" visible={this.props.sideBarVisibility} icon="labeled" vertical inverted>
          {renderMenuItem('home', 'Home', '/')}
          {renderMenuItem('library', 'Library', '/library')}
          {renderMenuItem('new journal', 'New Journal', '/')}
          <Menu.Item as="div" name="journal" onClick={this.props.toggleMenu}>
            Journal
            <Menu.Menu>
              <Link href="/journal" to="/journal">
                <Menu.Item>
                  Read
                </Menu.Item>
              </Link>
              <Link
                href="/newJournal"
                to="/newJournal"
              >
                <Menu.Item>
                  Create
                </Menu.Item>
              </Link>
            </Menu.Menu>
          </Menu.Item>
          { this.props.isAuthenticated
              ? this.renderLogout()
              : renderMenuItem('signin', 'Sign In', '/signin')
          }
        </Sidebar>
      </div>
    );
  }
}
