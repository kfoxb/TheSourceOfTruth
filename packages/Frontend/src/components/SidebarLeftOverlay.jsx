import React, { Component } from 'react';
import { Sidebar, Menu } from 'semantic-ui-react';
import PropTypes from 'prop-types';

export default class SidebarLeftOverlay extends Component {
  static propTypes = {
    sideBarVisibility: PropTypes.bool.isRequired,
    toggleMenu: PropTypes.func.isRequired,
  }

  componentDidMount() {
    document.addEventListener('mousedown', this.handleOffClick);
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleOffClick);
  }

  handleOffClick = (e) => {
    if (this.ref && !this.ref.contains(e.target) && this.props.sideBarVisibility) {
      this.props.toggleMenu();
    }
  };

  render() {
    return (
      <div ref={(el) => { this.ref = el; }}>
        <Sidebar as={Menu} animation="overlay" width="thin" visible={this.props.sideBarVisibility} icon="labeled" vertical inverted>
          <Menu.Item name="home">
            Home
          </Menu.Item>
          <Menu.Item name="humanityParty">
            Humanity Party
          </Menu.Item>
          <Menu.Item name="formatting">
            Formatting
          </Menu.Item>
          <Menu.Item name="transcription">
            Transcription
          </Menu.Item>
          <Menu.Item name="questions">
            Questions
          </Menu.Item>
          <Menu.Item name="journal">
            Journal
          </Menu.Item>
        </Sidebar>
      </div>
    );
  }
}
