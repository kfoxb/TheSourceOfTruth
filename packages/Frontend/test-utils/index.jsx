/* eslint-env jest */
/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { shallow } from 'enzyme';

export const createAssertWithPropsToMatchSnapshot = (Component, defaultProps) =>
  (extraProps) => {
    const props = { ...defaultProps, ...extraProps };
    const wrapper = shallow(<Component {...props} />);
    expect(wrapper).toMatchSnapshot();
  };
