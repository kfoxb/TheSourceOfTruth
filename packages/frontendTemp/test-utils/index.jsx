/* eslint-env jest */
/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { shallow } from 'enzyme';

export const makeShallowCreateWrapper = (Component, defaultProps) =>
  (extraProps) => {
    const props = { ...defaultProps, ...extraProps };
    return shallow(<Component {...props} />);
  };

export const createAssertWithPropsToMatchSnapshot = (Component, defaultProps) => {
  const createWrapper = makeShallowCreateWrapper(Component, defaultProps);
  return (extraProps) => {
    expect(createWrapper(extraProps)).toMatchSnapshot();
  };
};
