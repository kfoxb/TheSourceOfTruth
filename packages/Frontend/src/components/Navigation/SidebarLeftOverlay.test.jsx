/* eslint-env jest */
import React from 'react';
import { shallow } from 'enzyme';
import SidebarLeftOverlay from './SidebarLeftOverlay';
import { createAssertWithPropsToMatchSnapshot } from '../../../test-utils';

describe('SidebarLeftOverlay', () => {
  const noOp = () => {};
  const defaultProps = Object.freeze({
    isAuthenticated: false,
    logout: noOp,
    sideBarVisibility: true,
    toggleMenu: noOp,
  });
  describe('rendering', () => {
    const assertWithPropsToMatchSnapshot =
      createAssertWithPropsToMatchSnapshot(SidebarLeftOverlay, defaultProps);

    it('should render a sign in menu item when not authenticated', () => {
      assertWithPropsToMatchSnapshot();
    });

    it('should render a sign out menu item when authenticated', () => {
      assertWithPropsToMatchSnapshot({ isAuthenticated: true });
    });

    it('should not render a sidebar when sideBarVisibility is false', () => {
      assertWithPropsToMatchSnapshot({ sideBarVisibility: false });
    });

    it('should render a sidebar when sideBarVisibility is true', () => {
      assertWithPropsToMatchSnapshot({ sideBarVisibility: true });
    });
  });

  describe('event listeners', () => {
    it('should add an event listener on mousedown', () => {
      document.addEventListener = jest.fn();
      const inst = shallow(<SidebarLeftOverlay {...defaultProps} />).instance();
      expect(document.addEventListener).toHaveBeenCalledWith('mousedown', inst.handleOffClick);
    });

    it('should remove an event listener on mousedown when unmounted', () => {
      document.removeEventListener = jest.fn();
      const inst = shallow(<SidebarLeftOverlay {...defaultProps} />).instance();
      inst.componentWillUnmount();
      expect(document.removeEventListener).toHaveBeenCalledWith('mousedown', inst.handleOffClick);
    });
  });

  describe('handleOffClick', () => {
    it('should call toggleMenu, when it\'s ref does not contain the event target', () => {
      const toggleMenu = jest.fn();
      const inst = shallow(<SidebarLeftOverlay
        {...defaultProps}
        toggleMenu={toggleMenu}
      />).instance();
      const containingDiv = document.createElement('div');
      const outsideDiv = document.createElement('div');
      inst.ref = containingDiv;
      inst.handleOffClick({ target: outsideDiv });
      expect(toggleMenu).toHaveBeenCalled();
    });

    it('should not call toggleMenu, when it\'s ref contains the event target', () => {
      const toggleMenu = jest.fn();
      const inst = shallow(<SidebarLeftOverlay
        {...defaultProps}
        toggleMenu={toggleMenu}
      />).instance();
      const containingDiv = document.createElement('div');
      const childDiv = document.createElement('div');
      containingDiv.appendChild(childDiv);
      inst.ref = containingDiv;
      inst.handleOffClick({ target: childDiv });
      expect(toggleMenu).toHaveBeenCalledTimes(0);
    });
  });

  describe('handleRef', () => {
    it('should set this.ref to the passed in element', () => {
      const inst = shallow(<SidebarLeftOverlay {...defaultProps} />).instance();
      inst.handleRef('mock el');
      expect(inst.ref).toBe('mock el');
    });
  });

  describe('handleLogout', () => {
    it('should be a method on SidebarLeftOverlay', () => {
      const inst = shallow(<SidebarLeftOverlay {...defaultProps} />).instance();
      expect(typeof inst.handleLogout).toBe('function');
    });

    it('should call props.logout', () => {
      const logout = jest.fn();
      const inst = shallow(<SidebarLeftOverlay
        {...defaultProps}
        logout={logout}
      />).instance();
      inst.handleLogout();
      expect(logout).toHaveBeenCalledTimes(1);
    });

    it('should call props.toggleMenu', () => {
      const toggleMenu = jest.fn();
      const inst = shallow(<SidebarLeftOverlay
        {...defaultProps}
        toggleMenu={toggleMenu}
      />).instance();
      inst.handleLogout();
      expect(toggleMenu).toHaveBeenCalledTimes(1);
    });
  });
});
