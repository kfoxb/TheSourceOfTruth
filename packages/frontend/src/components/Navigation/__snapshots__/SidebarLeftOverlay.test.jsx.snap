// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarLeftOverlay rendering should not render a sidebar when sideBarVisibility is false 1`] = `
<div>
  <Sidebar
    animation="overlay"
    as={[Function]}
    direction="left"
    icon="labeled"
    inverted={true}
    vertical={true}
    visible={false}
    width="thin"
  >
    <Link
      href="/eng/"
      replace={false}
      to="/eng/"
    >
      <MenuItem
        as="div"
        name="home"
        onClick={[Function]}
      >
        Home
      </MenuItem>
    </Link>
    <Link
      href="/eng/library"
      replace={false}
      to="/eng/library"
    >
      <MenuItem
        as="div"
        name="library"
        onClick={[Function]}
      >
        Library
      </MenuItem>
    </Link>
    <Link
      href="/eng/tasks"
      replace={false}
      to="/eng/tasks"
    >
      <MenuItem
        as="div"
        name="tasks"
        onClick={[Function]}
      >
        Tasks
      </MenuItem>
    </Link>
    <Link
      href="/eng/signin"
      replace={false}
      to="/eng/signin"
    >
      <MenuItem
        as="div"
        name="signin"
        onClick={[Function]}
      >
        Sign In
      </MenuItem>
    </Link>
  </Sidebar>
</div>
`;

exports[`SidebarLeftOverlay rendering should render a sidebar when sideBarVisibility is true 1`] = `
<div>
  <Sidebar
    animation="overlay"
    as={[Function]}
    direction="left"
    icon="labeled"
    inverted={true}
    vertical={true}
    visible={true}
    width="thin"
  >
    <Link
      href="/eng/"
      replace={false}
      to="/eng/"
    >
      <MenuItem
        as="div"
        name="home"
        onClick={[Function]}
      >
        Home
      </MenuItem>
    </Link>
    <Link
      href="/eng/library"
      replace={false}
      to="/eng/library"
    >
      <MenuItem
        as="div"
        name="library"
        onClick={[Function]}
      >
        Library
      </MenuItem>
    </Link>
    <Link
      href="/eng/tasks"
      replace={false}
      to="/eng/tasks"
    >
      <MenuItem
        as="div"
        name="tasks"
        onClick={[Function]}
      >
        Tasks
      </MenuItem>
    </Link>
    <Link
      href="/eng/signin"
      replace={false}
      to="/eng/signin"
    >
      <MenuItem
        as="div"
        name="signin"
        onClick={[Function]}
      >
        Sign In
      </MenuItem>
    </Link>
  </Sidebar>
</div>
`;

exports[`SidebarLeftOverlay rendering should render a sign in menu item when not authenticated 1`] = `
<div>
  <Sidebar
    animation="overlay"
    as={[Function]}
    direction="left"
    icon="labeled"
    inverted={true}
    vertical={true}
    visible={true}
    width="thin"
  >
    <Link
      href="/eng/"
      replace={false}
      to="/eng/"
    >
      <MenuItem
        as="div"
        name="home"
        onClick={[Function]}
      >
        Home
      </MenuItem>
    </Link>
    <Link
      href="/eng/library"
      replace={false}
      to="/eng/library"
    >
      <MenuItem
        as="div"
        name="library"
        onClick={[Function]}
      >
        Library
      </MenuItem>
    </Link>
    <Link
      href="/eng/tasks"
      replace={false}
      to="/eng/tasks"
    >
      <MenuItem
        as="div"
        name="tasks"
        onClick={[Function]}
      >
        Tasks
      </MenuItem>
    </Link>
    <Link
      href="/eng/signin"
      replace={false}
      to="/eng/signin"
    >
      <MenuItem
        as="div"
        name="signin"
        onClick={[Function]}
      >
        Sign In
      </MenuItem>
    </Link>
  </Sidebar>
</div>
`;

exports[`SidebarLeftOverlay rendering should render a sign out menu item when authenticated non-anonymously 1`] = `
<div>
  <Sidebar
    animation="overlay"
    as={[Function]}
    direction="left"
    icon="labeled"
    inverted={true}
    vertical={true}
    visible={true}
    width="thin"
  >
    <Link
      href="/eng/"
      replace={false}
      to="/eng/"
    >
      <MenuItem
        as="div"
        name="home"
        onClick={[Function]}
      >
        Home
      </MenuItem>
    </Link>
    <Link
      href="/eng/library"
      replace={false}
      to="/eng/library"
    >
      <MenuItem
        as="div"
        name="library"
        onClick={[Function]}
      >
        Library
      </MenuItem>
    </Link>
    <Link
      href="/eng/tasks"
      replace={false}
      to="/eng/tasks"
    >
      <MenuItem
        as="div"
        name="tasks"
        onClick={[Function]}
      >
        Tasks
      </MenuItem>
    </Link>
    <MenuItem
      as="div"
      name="logout"
      onClick={[Function]}
    >
      Sign Out
    </MenuItem>
  </Sidebar>
</div>
`;
